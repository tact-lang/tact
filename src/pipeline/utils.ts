import type { TypeDescription } from "../types/types";

/**
 * This function sorts contracts in topological order.
 * It also checks for cycles in the graph of dependencies and returns undefined if a cycle is found.
 * It works in O(N) time, where N is the number of contracts.
 * @param allContracts - list of all contracts.
 * @returns sorted list of contracts or undefined if a cycle is found
 */
export const topSortContracts = (
    allContracts: TypeDescription[],
): TypeDescription[] | undefined => {
    const visitingNow: Set<TypeDescription> = new Set();
    const visited: Set<TypeDescription> = new Set();
    const result: TypeDescription[] = [];
    const dfs = (contract: TypeDescription): boolean => {
        visitingNow.add(contract);
        for (const c of contract.dependsOn) {
            if (c.kind !== "contract") {
                // This check should also be false. As contracts should be filtered out before calling this function.
                continue;
            }
            if (visitingNow.has(c)) {
                return false;
            }
            if (!visited.has(c)) {
                if (!dfs(c)) {
                    return false;
                }
            }
        }
        visitingNow.delete(contract);
        visited.add(contract);
        result.push(contract);
        return true;
    };
    for (const contract of allContracts) {
        if (!visited.has(contract)) {
            if (!dfs(contract)) {
                return undefined;
            }
        }
    }
    return result;
};
