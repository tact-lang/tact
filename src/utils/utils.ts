import { crc16 } from "./crc16";

export function topologicalSort<T>(src: T[], references: (src: T) => T[]) {
    const result: T[] = [];
    const visited = new Set<T>();
    const visiting = new Set<T>();
    const visit = (src: T) => {
        if (visiting.has(src)) {
            throw Error('Cycle detected');
        }
        if (!visited.has(src)) {
            visiting.add(src);
            for (const r of references(src)) {
                visit(r);
            }
            visiting.delete(src);
            visited.add(src);
            result.push(src);
        }
    }
    for (const s of src) {
        visit(s);
    }
    return result;
}

export function deepFreeze<T>(obj: T) {
    const propNames = Object.getOwnPropertyNames(obj);
    for (const name of propNames) {
        // eslint-disable-next-line @typescript-eslint/no-explicit-any
        const value = (obj as any)[name];
        if (value && typeof value === "object") {
            deepFreeze(value);
        }
    }
    return Object.freeze(obj);
}

export function getMethodId(name: string) {
    return (crc16(name) & 0xffff) | 0x10000;
}