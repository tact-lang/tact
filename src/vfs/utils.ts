import { sha256_sync } from "@ton/crypto";
import path from "path";

/**
 * Ensures the resulting file name does not exceed the given maximum length.
 * If too long, trims the name and appends a short hash to avoid collisions.
 *
 * @param name - The base file name without extension.
 * @param ext - The file extension.
 * @param maxLen - Maximum allowed length for the full file name (default: 255).
 * @returns A safe file name within the specified length.
 */
export const makeSafeName = (
    name: string,
    ext: string,
    maxLen = 255,
): string => {
    const full = name + ext;

    if (full.length <= maxLen) {
        return full;
    }

    const hash = sha256_sync(Buffer.from(name)).toString("hex").slice(0, 8);
    const suffix = `_${hash}${ext}`;
    const maxNameLen = maxLen - suffix.length;
    const safeName = name.slice(0, maxNameLen);

    return `${safeName}${suffix}`;
};

/**
 * Returns the full extension of a file, including all parts after the first dot.
 * - "file.txt" => ".txt"
 * - "archive.tar.gz" => ".tar.gz"
 */
export const getFullExtension = (filename: string): string => {
    const base = path.basename(filename);
    const firstDotIndex = base.indexOf(".");
    return firstDotIndex !== -1 ? base.slice(firstDotIndex) : "";
};
