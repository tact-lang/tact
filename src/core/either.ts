// left$0 {X:Type} {Y:Type} value:X = Either X Y;
// right$1 {X:Type} {Y:Type} value:Y = Either X Y;

import { uint } from "@/core/numeric";
import type { Tlb } from "@/core/tlb";

export type Either<L, R> = Left<L> | Right<R>
export type Left<L> = { readonly $: 'left', readonly value: L }
export const Left = <L>(value: L): Left<L> => Object.freeze({ $: 'left', value });
export const isLeft = <L>(m: Either<L, unknown>): m is Left<L> => m.$ === 'left';
export type Right<R> = { readonly $: 'right', readonly value: R }
export const Right = <R>(value: R): Right<R> => Object.freeze({ $: 'right', value });
export const isRight = <R>(m: Either<unknown, R>): m is Right<R> => m.$ === 'right';

export const either = <L, R>(left: Tlb<L>, right: Tlb<R>): Tlb<Either<L, R>> => ({
    store: (t, b, p) => {
        if (t.$ === 'left') {
            uint(1).store(0, b, [...p, '$']);
            left.store(t.value, b, [...p, 'value']);
        } else {
            uint(1).store(1, b, [...p, '$']);
            right.store(t.value, b, [...p, 'value']);
        }
    },
    load: (s, p) => {
        const type = uint(1).load(s, [...p, '$']);
        if (type === 0) {
            return Left(left.load(s, [...p, 'value']));
        } else {
            return Right(right.load(s, [...p, 'value']));
        }
    },
});
