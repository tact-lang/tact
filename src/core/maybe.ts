// nothing$0 {X:Type} = Maybe X;
// just$1 {X:Type} value:X = Maybe X;

import { TlbError, type Path, type Tlb } from "@/core/tlb";
import type { Builder, Slice } from "@ton/core";

export type Maybe<T> = Nothing | Just<T>
export type Nothing = { readonly $: 'nothing' }
export const Nothing: Nothing = Object.freeze({ $: 'nothing' })
export const isNothing = <T>(m: Maybe<T>): m is Nothing => m.$ === 'nothing';
export type Just<T> = { readonly $: 'just', readonly value: T }
export const Just = <T>(value: T): Just<T> => Object.freeze({ $: 'just', value });
export const isJust = <T>(m: Maybe<T>): m is Just<T> => m.$ === 'just';

export const maybe = <T>(type: Tlb<T>): Tlb<Maybe<T>> => {
    const store = (t: Maybe<T>, b: Builder, p: Path) => {
        if (t.$ === 'nothing') {
            b.storeBit(0);
            return;
        }
        try {
            b.storeBit(1);
        } catch (e) {
            throw new TlbError(e, [...p, '$']);
        }
        type.store(t.value, b, [...p, 'value']);
    };
    const load = (s: Slice, p: Path) => {
        const bit = (() => {
            try {
                return s.loadBit();
            } catch (e) {
                throw new TlbError(e, [...p, '$']);
            }
        })();
        if (bit) {
            return Just(type.load(s, [...p, 'value']));
        } else {
            return Nothing;
        }
    };
    return { store, load };
};