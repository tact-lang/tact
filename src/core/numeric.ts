import { _primitive, type Tlb } from "@/core/tlb";

export const bit: Tlb<boolean> = _primitive({
    store: (t, b) => b.storeBit(t),
    load: (s) => s.loadBit(),
});

export const int = (bits: number): Tlb<number> => _primitive({
    store: (t, b) => b.storeInt(t, bits),
    load: (s) => s.loadInt(bits),
});

export const uint = (bits: number): Tlb<number> => _primitive({
    store: (t, b) => b.storeUint(t, bits),
    load: (s) => s.loadUint(bits),
});

export const intVar = (bits: number): Tlb<number> => _primitive({
    store: (t, b) => b.storeVarInt(t, bits),
    load: (s) => s.loadVarInt(bits),
});

export const uintVar = (bits: number): Tlb<number> => _primitive({
    store: (t, b) => b.storeVarUint(t, bits),
    load: (s) => s.loadVarUint(bits),
});

export const coins = uintVar(16);

export const intBig = (bits: number): Tlb<bigint> => _primitive({
    store: (t, b) => b.storeInt(t, bits),
    load: (s) => s.loadIntBig(bits),
});

export const uintBig = (bits: number): Tlb<bigint> => _primitive({
    store: (t, b) => b.storeUint(t, bits),
    load: (s) => s.loadUintBig(bits),
});

export const intVarBig = (bits: number): Tlb<bigint> => _primitive({
    store: (t, b) => b.storeVarInt(t, bits),
    load: (s) => s.loadVarIntBig(bits),
});

export const uintVarBig = (bits: number): Tlb<bigint> => _primitive({
    store: (t, b) => b.storeVarUint(t, bits),
    load: (s) => s.loadVarUintBig(bits),
});