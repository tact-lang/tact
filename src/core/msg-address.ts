// addr_none$00 = MsgAddressExt;
// addr_extern$01 len:(## 9) external_address:(bits len) = MsgAddressExt;
// anycast_info$_ depth:(#<= 30) { depth >= 1 }
//    rewrite_pfx:(bits depth) = Anycast;
// addr_std$10 anycast:(Maybe Anycast) workchain_id:int8 address:bits256  = MsgAddressInt;
// addr_var$11 anycast:(Maybe Anycast) addr_len:(## 9) workchain_id:int32 address:(bits addr_len) = MsgAddressInt;
// _ _:MsgAddressInt = MsgAddress;
// _ _:MsgAddressExt = MsgAddress;

import { int, uint, uintBig } from "@/core/numeric";
import type { Path, Tlb } from "@/core/tlb";
import type { Builder, Slice } from "@ton/core";

export type MsgAddress =
    | AddrNone
    | AddrExtern
    | AddrStd
    | AddrVar
export type AddrNone = {
    readonly $: 'AddrNone'
}
export const AddrNone: AddrNone = Object.freeze({ $: "AddrNone" });
export const isAddrNone = (a: MsgAddress): a is AddrNone => a.$ === 'AddrNone';
export type AddrExtern = {
    readonly $: 'AddrExtern'
    readonly len: number
    readonly address: bigint; // bits len
}
export const AddrExtern = (len: number, address: bigint): AddrExtern => Object.freeze({
    $: 'AddrExtern', address, len,
})
export const isAddrExtern = (a: MsgAddress): a is AddrExtern => a.$ === 'AddrExtern';
export type AddrStd = {
    readonly $: 'AddrStd'
    readonly workchainId: number;
    readonly address: bigint;
}
export const AddrStd = (workchainId: number, address: bigint): AddrStd => Object.freeze({
    $: 'AddrStd', address, workchainId,
})
export const isAddrStd = (a: MsgAddress): a is AddrStd => a.$ === 'AddrStd';
export type AddrVar = {
    readonly $: 'AddrVar'
    readonly len: number // ## 9
    readonly workchainId: number // int32
    readonly address: bigint // bits addr_len
}
export const AddrVar = (len: number, workchainId: number, address: bigint): AddrVar => Object.freeze({
    $: "AddrVar", address, len, workchainId,
});
export const isAddrVar = (a: MsgAddress): a is AddrVar => a.$ === 'AddrVar';

const loadAddress = (s: Slice, p: Path): MsgAddress => {
    const type = uint(2).load(s, [...p, '$']);
    if (type === 0) {
        // addr_none$00
        return AddrNone;
    } else if (type === 1) {
        // addr_extern$01
        // len:(## 9)
        const len = uint(9).load(s, [...p, 'len']);
        // external_address:(bits len)
        const addr = uintBig(len).load(s, [...p, 'address']);
        return AddrExtern(len, addr);
    } else if (type === 2) {
        // addr_std$10
        // anycast:(Maybe Anycast)
        if (uint(1).load(s, [...p, 'anycast']) !== 0) {
            throw Error('Anycast not supported');
        }
        // workchain_id:int8
        const wc = int(8).load(s, [...p, 'workchainId']);
        // address:bits256
        const addr = uintBig(256).load(s, [...p, 'address']);
        return AddrStd(wc, addr);
    } else if (type === 3) {
        // addr_var$11
        // anycast:(Maybe Anycast)
        if (uint(1).load(s, [...p, 'anycast']) !== 0) {
            throw Error('Anycast not supported');
        }
        // addr_len:(## 9)
        const len = uint(9).load(s, [...p, 'len']);
        // workchain_id:int32
        const wc = int(32).load(s, [...p, 'workchainId']);
        // address:(bits addr_len)
        const addr = uintBig(len).load(s, [...p, 'address']);
        return AddrVar(len, wc, addr);
    } else {
        throw Error('Unreachable');
    }
};

const storeAddress = (t: MsgAddress, b: Builder, p: Path) => {
    switch (t.$) {
        case 'AddrNone': {
            // addr_none$00
            uint(2).store(0, b, [...p, '$']);
            return;
        }
        case 'AddrExtern': {
            // addr_extern$01
            uint(2).store(1, b, [...p, '$']);
            // len:(## 9)
            uint(9).store(t.len, b, [...p, 'len']);
            // external_address:(bits len)
            uintBig(t.len).store(t.address, b, [...p, 'address']);
            return;
        }
        case 'AddrStd': {
            // addr_std$10
            uint(2).store(2, b, [...p, '$']);
            // anycast:(Maybe Anycast)
            uint(1).store(0, b, [...p, 'anycast']);
            // workchain_id:int8
            int(8).store(t.workchainId, b, [...p, 'workchainId']);
            // address:bits256
            uintBig(256).store(t.address, b, [...p, 'address']);
            return;
        }
        case 'AddrVar': {
            // addr_var$11
            uint(2).store(3, b, [...p, '$']);
            // anycast:(Maybe Anycast)
            uint(1).store(0, b, [...p, 'anycast']);
            // addr_len:(## 9)
            uint(9).store(t.len, b, [...p, 'len']);
            // workchain_id:int32
            int(32).store(t.workchainId, b, [...p, 'workchainId']);
            // address:(bits addr_len)
            uintBig(t.len).store(t.address, b, [...p, 'address']);
        }
    }
};

export const msgAddress: Tlb<MsgAddress> = ({
    store: storeAddress,
    load: loadAddress,
});