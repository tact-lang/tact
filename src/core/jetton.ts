// InternalMsgBody
//     transfer
//         #0f8a7ea5
//         query_id:uint64
//         amount:(VarUInteger 16)
//         destination:MsgAddress
//         response_destination:MsgAddress
//         custom_payload:(Maybe ^Cell)
//         forward_ton_amount:(VarUInteger 16)
//         forward_payload:(Either Cell ^Cell)
//     transfer_notification
//         #7362d09c
//         query_id:uint64
//         amount:(VarUInteger 16)
//         sender:MsgAddress
//         forward_payload:(Either Cell ^Cell)
//     excesses
//         #d53276db
//         query_id:uint64
//     burn
//         #595f07bc
//         query_id:uint64
//         amount:(VarUInteger 16)
//         response_destination:MsgAddress
//         custom_payload:(Maybe ^Cell)
//     internal_transfer
//         #178d4519
//         query_id:uint64
//         amount:(VarUInteger 16)
//         from:MsgAddress
//         response_address:MsgAddress
//         forward_ton_amount:(VarUInteger 16)
//         forward_payload:(Either Cell ^Cell)
//     burn_notification
//         #7bdd97de
//         query_id:uint64
//         amount:(VarUInteger 16)
//         sender:MsgAddress
//         response_destination:MsgAddress

import { either } from "@/core/either";
import { maybe } from "@/core/maybe";
import { msgAddress } from "@/core/msg-address";
import { uint, uintVar } from "@/core/numeric";
import { type GetTlb, literal, object, ref, type Tlb } from "@/core/tlb";

export const transfer = <Custom, Forward>(
    custom: Tlb<Custom>,
    forward: Tlb<Forward>,
) => {
    return object({
        $: literal(uint(32))(0x0f8a7ea5),
        queryId: uint(64),
        amount: uintVar(16),
        destination: msgAddress,
        responseDestination: msgAddress,
        customPayload: maybe(ref(custom)),
        forwardTonAmount: uintVar(16),
        forwardPayload: either(forward, ref(forward)),
    });
};

export type Transfer<T, U> = GetTlb<ReturnType<typeof transfer<T, U>>>
