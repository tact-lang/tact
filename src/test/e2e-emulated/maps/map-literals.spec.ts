import "@ton/test-utils";

import { toNano } from "@ton/core";
import { Blockchain } from "@ton/sandbox";
import { Flat } from "./output/map-literals_Flat";

import { zeroStoragePrices } from "@/test/utils/gasUtils";

const deployValue = toNano("1"); // `dump` is expensive

const setup = async () => {
    const blockchain = await Blockchain.create();
    blockchain.verbosity.print = false;

    blockchain.setConfig(zeroStoragePrices(blockchain.config));

    const treasury = await blockchain.treasury("treasury");

    const contract = blockchain.openContract(await Flat.fromInit());

    const deployResult = await contract.send(
        treasury.getSender(),
        { value: deployValue },
        null,
    );
    expect(deployResult.transactions).toHaveTransaction({
        from: treasury.address,
        to: contract.address,
        success: true,
        deploy: true,
    });

    return {
        blockchain,
        treasury,
        contract,
    };
};

it("should correctly compile in functions", async () => {
    const { contract } = await setup();
    const result = await contract.getInState();
    expect(result).toMatchSnapshot();
});

it("should correctly compile empty literals", async () => {
    const { contract } = await setup();
    const result = await contract.getEmpty();
    expect(result).toMatchSnapshot();
});

it("should correctly compile const in contract", async () => {
    const { contract } = await setup();
    const result = await contract.getContractConst();
    expect(result).toMatchSnapshot();
});

it("should correctly compile global const", async () => {
    const { contract } = await setup();
    const result = await contract.getGlobalConst();
    expect(result).toMatchSnapshot();
});

it("should create maps with address keys", async () => {
    const { contract } = await setup();
    const result = await contract.getAddress();
    expect(result).toMatchSnapshot();
});

it("should work with large maps", async () => {
    const { contract } = await setup();
    const result = await contract.getLarge();
    expect(result).toMatchSnapshot();
});

it("should match explicitly constructed maps", async () => {
    const { contract } = await setup();
    const result = await contract.getEq();
    expect(result).toBeTruthy();
});

it("should work with expressions in keys and values", async () => {
    const { contract } = await setup();
    const result = await contract.getExpression();
    expect(result).toMatchSnapshot();
});

it("should keep get/set invariants of regular maps", async () => {
    const { contract } = await setup();
    const result = await contract.getGetSetResult();
    expect(result).toBeTruthy();
});

it("should work as global constant map", async () => {
    const { contract } = await setup();
    const result = await contract.getDecode();
    expect(result).toMatchSnapshot();
});

it("should work anonymously", async () => {
    const { contract } = await setup();
    const result = await contract.getAnonymous();
    expect(result).toMatchSnapshot();
});

it("the last value of the same key should be used", async () => {
    const { contract } = await setup();
    const result = await contract.getDuplicates();
    expect(result).toMatchSnapshot();
});
