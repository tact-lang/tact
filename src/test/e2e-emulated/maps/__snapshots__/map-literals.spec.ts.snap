// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should correctly compile const in contract 1`] = `
Dictionary {
  "_key": {
    "bits": 16,
    "parse": [Function],
    "serialize": [Function],
  },
  "_map": Map {
    "n:-1" => 2,
  },
  "_value": {
    "parse": [Function],
    "serialize": [Function],
  },
}
`;

exports[`should correctly compile empty literals 1`] = `
Dictionary {
  "_key": {
    "bits": 16,
    "parse": [Function],
    "serialize": [Function],
  },
  "_map": Map {},
  "_value": {
    "parse": [Function],
    "serialize": [Function],
  },
}
`;

exports[`should correctly compile global const 1`] = `
Dictionary {
  "_key": {
    "bits": 16,
    "parse": [Function],
    "serialize": [Function],
  },
  "_map": Map {
    "n:1" => -2,
  },
  "_value": {
    "parse": [Function],
    "serialize": [Function],
  },
}
`;

exports[`should correctly compile in functions 1`] = `
Dictionary {
  "_key": {
    "bits": 16,
    "parse": [Function],
    "serialize": [Function],
  },
  "_map": Map {
    "n:1234" => 2345,
    "n:8765" => 7654,
  },
  "_value": {
    "parse": [Function],
    "serialize": [Function],
  },
}
`;

exports[`should create maps with address keys 1`] = `
Dictionary {
  "_key": {
    "bits": 267,
    "parse": [Function],
    "serialize": [Function],
  },
  "_map": Map {
    "a:EQAYqo4u7VF0fa4DPAebk4g9lBytj2VFny7pzXR0trjtXQaO" => 2000000000n,
    "a:EQDtFpEwcFAEcRe5mLVh2N6C0x-_hJEM7W61_JLnSF74p4q2" => 1000000000n,
  },
  "_value": {
    "parse": [Function],
    "serialize": [Function],
  },
}
`;

exports[`should work anonymously 1`] = `3n`;

exports[`should work as global constant map 1`] = `-2n`;

exports[`should work with expressions in keys and values 1`] = `
Dictionary {
  "_key": {
    "bits": 16,
    "parse": [Function],
    "serialize": [Function],
  },
  "_map": Map {
    "n:3" => 7,
  },
  "_value": {
    "parse": [Function],
    "serialize": [Function],
  },
}
`;

exports[`should work with large maps 1`] = `
Dictionary {
  "_key": {
    "bits": 257,
    "parse": [Function],
    "serialize": [Function],
  },
  "_map": Map {
    "b:0" => 0n,
    "b:1" => 1n,
    "b:2" => 2n,
    "b:3" => 3n,
    "b:4" => 4n,
    "b:5" => 5n,
    "b:6" => 6n,
    "b:7" => 7n,
    "b:8" => 8n,
    "b:9" => 9n,
    "b:10" => 10n,
    "b:11" => 11n,
    "b:12" => 12n,
    "b:13" => 13n,
    "b:14" => 14n,
    "b:15" => 15n,
    "b:16" => 16n,
    "b:17" => 17n,
    "b:18" => 18n,
    "b:19" => 19n,
    "b:20" => 20n,
    "b:21" => 21n,
    "b:22" => 22n,
    "b:23" => 23n,
    "b:24" => 24n,
    "b:25" => 25n,
    "b:26" => 26n,
    "b:27" => 27n,
    "b:28" => 28n,
    "b:29" => 29n,
    "b:30" => 30n,
    "b:31" => 31n,
    "b:32" => 32n,
    "b:33" => 33n,
    "b:34" => 34n,
    "b:35" => 35n,
    "b:36" => 36n,
    "b:37" => 37n,
    "b:38" => 38n,
    "b:39" => 39n,
    "b:40" => 40n,
    "b:41" => 41n,
    "b:42" => 42n,
    "b:43" => 43n,
    "b:44" => 44n,
    "b:45" => 45n,
    "b:46" => 46n,
    "b:47" => 47n,
    "b:48" => 48n,
    "b:49" => 49n,
    "b:50" => 50n,
    "b:51" => 51n,
    "b:52" => 52n,
    "b:53" => 53n,
    "b:54" => 54n,
    "b:55" => 55n,
    "b:56" => 56n,
    "b:57" => 57n,
    "b:58" => 58n,
    "b:59" => 59n,
    "b:60" => 60n,
    "b:61" => 61n,
    "b:62" => 62n,
    "b:63" => 63n,
    "b:64" => 64n,
    "b:65" => 65n,
    "b:66" => 66n,
    "b:67" => 67n,
    "b:68" => 68n,
    "b:69" => 69n,
    "b:70" => 70n,
    "b:71" => 71n,
    "b:72" => 72n,
    "b:73" => 73n,
    "b:74" => 74n,
    "b:75" => 75n,
    "b:76" => 76n,
    "b:77" => 77n,
    "b:78" => 78n,
    "b:79" => 79n,
    "b:80" => 80n,
    "b:81" => 81n,
    "b:82" => 82n,
    "b:83" => 83n,
    "b:84" => 84n,
    "b:85" => 85n,
    "b:86" => 86n,
    "b:87" => 87n,
    "b:88" => 88n,
    "b:89" => 89n,
    "b:90" => 90n,
    "b:91" => 91n,
    "b:92" => 92n,
    "b:93" => 93n,
    "b:94" => 94n,
    "b:95" => 95n,
    "b:96" => 96n,
    "b:97" => 97n,
    "b:98" => 98n,
    "b:99" => 99n,
    "b:100" => 100n,
    "b:101" => 101n,
    "b:102" => 102n,
    "b:103" => 103n,
    "b:104" => 104n,
    "b:105" => 105n,
    "b:106" => 106n,
    "b:107" => 107n,
    "b:108" => 108n,
    "b:109" => 109n,
    "b:110" => 110n,
    "b:111" => 111n,
    "b:112" => 112n,
    "b:113" => 113n,
    "b:114" => 114n,
    "b:115" => 115n,
    "b:116" => 116n,
    "b:117" => 117n,
    "b:118" => 118n,
    "b:119" => 119n,
    "b:120" => 120n,
    "b:121" => 121n,
    "b:122" => 122n,
    "b:123" => 123n,
    "b:124" => 124n,
    "b:125" => 125n,
    "b:126" => 126n,
    "b:127" => 127n,
    "b:128" => 128n,
    "b:129" => 129n,
    "b:130" => 130n,
    "b:131" => 131n,
    "b:132" => 132n,
    "b:133" => 133n,
    "b:134" => 134n,
    "b:135" => 135n,
    "b:136" => 136n,
    "b:137" => 137n,
    "b:138" => 138n,
    "b:139" => 139n,
    "b:140" => 140n,
    "b:141" => 141n,
    "b:142" => 142n,
    "b:143" => 143n,
    "b:144" => 144n,
    "b:145" => 145n,
    "b:146" => 146n,
    "b:147" => 147n,
    "b:148" => 148n,
    "b:149" => 149n,
    "b:150" => 150n,
    "b:151" => 151n,
    "b:152" => 152n,
    "b:153" => 153n,
    "b:154" => 154n,
    "b:155" => 155n,
    "b:156" => 156n,
    "b:157" => 157n,
    "b:158" => 158n,
    "b:159" => 159n,
    "b:160" => 160n,
    "b:161" => 161n,
    "b:162" => 162n,
    "b:163" => 163n,
    "b:164" => 164n,
    "b:165" => 165n,
    "b:166" => 166n,
    "b:167" => 167n,
    "b:168" => 168n,
    "b:169" => 169n,
    "b:170" => 170n,
    "b:171" => 171n,
    "b:172" => 172n,
    "b:173" => 173n,
    "b:174" => 174n,
    "b:175" => 175n,
    "b:176" => 176n,
    "b:177" => 177n,
    "b:178" => 178n,
    "b:179" => 179n,
    "b:180" => 180n,
    "b:181" => 181n,
    "b:182" => 182n,
    "b:183" => 183n,
    "b:184" => 184n,
    "b:185" => 185n,
    "b:186" => 186n,
    "b:187" => 187n,
    "b:188" => 188n,
    "b:189" => 189n,
    "b:190" => 190n,
    "b:191" => 191n,
    "b:192" => 192n,
    "b:193" => 193n,
    "b:194" => 194n,
    "b:195" => 195n,
    "b:196" => 196n,
    "b:197" => 197n,
    "b:198" => 198n,
    "b:199" => 199n,
  },
  "_value": {
    "parse": [Function],
    "serialize": [Function],
  },
}
`;

exports[`the last value of the same key should be used 1`] = `4n`;
